---
title: "Use mirTarRnaSeq for miRNA mRNA interaction analysis"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{mirTarRnaSeq}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
library(knitr)
library(rmarkdown)
library(mirTarRnaSeq)
library(dplyr)

doctype <- opts_knit$get("rmarkdown.pandoc.to")
```

# Introduction

mirTarRnaSeq is a package for miRNA and mRNA interaction analysis through regression and correlation approaches. miRNA and mRNA seq data are analysed from the same experiment (condition or time point data) and mRNA targets of the miRNAs from same experiment will be identified using statistical approaches.

The example data set for the first approach is 8 matching miRNA and mRNA EBV positive samples from TCGA. We attempt to identify the EBV miRNA targets on both human. 

The second example set set is the mouse fibroblast differntiated to muscle cells in three time points. Here, we try to identify mRNA targets of miRNA experssed at various time points.

## Data upload

### mirTarRnaSeq accepts data in dataframe or table formats

* For the first approach ([Part1](#part1-uploading-data-into-the-application)) we use a table of differntially expressed mRNA genes in human from a high-low EBV differential expression experiment. Note that we are choosing these differentially expressed genes to decrease the number of our potential targets. The user might choose to use all expressed gene but the analysis will be more time consuming and open to the biological question interpertation.
* Next we use a list of normalized (tpm) EBV miRNA experession data from the same samples. 
* For the second part (Part2, Part3) of the experiment we are using two tables of differentially expressed mRNA and miRNA sequencing fold change results from mouse time point specific differntiation experiments.

# Part1: Uploading data into the application

The example data can be found in the test folder under package.

```{r eval=TRUE, echo=TRUE}
DiffExp<-read.table("test/HumanSigGenesH_L.txt", as.is=TRUE, header=T, row.names=1)
miRNAExp<-read.table("test/m_Stom_data_auto_miRNA.txt", as.is=TRUE, header=T, row.names=1)
```

## Transpose data

We will transpose the data so the miRNA and mRNA data can be comparable and scalable using the `tzTrans` function

```{r eval=TRUE, echo=TRUE}
DiffExpmiRNA <- tzTrans(miRNAExp)
DiffExpmRNA <- tzTrans(DiffExp)

```

## Get miranda file

We currently support miranda runs (potential miRNA target parsing by score,interaction energy, and interaction or miRNA length ) on seven species ("Human", "Mouse", "Drosophila", "C.elegans", "Epstein_Barr" (EBV), "Cytomegalovirus" (CMV) and "Kaposi_Sarcoma" (KSHV)). We also support the viral miRNAs targeting human genes for EBV "Epstein_Barr_Human", CMV "CMV_Human" and KSHV "KSHV_Human". 
1) Here we first import the relevant miranda file .
2) We only keep targets which are also targets of EBV miRNAs based on our EBV miranda file.

```{r eval=TRUE, echo=TRUE}
miranda <- getInputSpecies("Epstein_Barr_Human", threshold = 180)
DiffExpmRNASub <- miRanComp(DiffExpmRNA, miranda)
```

## Select miRNA

```{r eval=TRUE, echo=TRUE}
miRNA_select<-c("ebv-mir-BART8-5p")
```

## Combine the mRNA and miRNA file and define boundaries

```{r eval=TRUE, echo=TRUE}
Combine<- combiner(DiffExpmRNA,DiffExpmiRNA,miRNA_select)
geneVariant<- geneVari(Combine,miRNA_select)
```

## Regression model run

```{r eval=TRUE, echo=TRUE}
MRun<- runModels(Combine,geneVariant,miRNA_select,mode="multi")
```

## FDR sig Genes

```{r eval=TRUE, echo=TRUE}
FDRModel<- fdrSig(MRun, value=0.1,method="fdr")
SigFDRGenes<-as.data.frame(cbind(FDRModel$pvalues,FDRModel$FDR_Value))
names(SigFDRGenes)<-c("P Value","FDR")
```

## Including Plots

```{r eval=TRUE, echo=TRUE, fig.width = 5, fig.height=5, out.width="80%", dpi=300, fig.align="center"}
mymodel <- FDRModel$all_models[[1]]  # pick the first
plotFit(mymodel)
plotResiduals(mymodel)
plotTerms(mymodel)
```

## Run a vector of miRNAs with "inter" or "multi" mode

```{r eval=TRUE, echo=TRUE}
miRNA_select<-c("ebv-mir-BART17-3p","ebv-mir-BART2")
Combine<- combiner(DiffExpmRNA,DiffExpmiRNA,miRNA_select)
geneVariant<- geneVari(Combine,miRNA_select)
MRun<- runModels(Combine,geneVariant,miRNA_select,mode="multi")
FDRModel<-fdrSig(MRun, value=0.1,method="fdr")
table(FDRModel$FDR_significant)
SigFDRGenes<-as.data.frame(cbind(FDRModel$pvalues,FDRModel$FDR_Value))
names(SigFDRGenes)<-c("P Value","FDR")
```

## Including Plots

```{r eval=TRUE, echo=TRUE, fig.width = 5, fig.height=5, out.width="80%", dpi=300, fig.align="center"}
mymodel <- FDRModel$all_models[[1]]  # pick the first
plotFit(mymodel)
plotResiduals(mymodel)
plotTerms(mymodel)
```

## Run for all miRNAs

Mode can be  "inter", "multi" or "NULL" mode
Get unique miRNA names which are also present in your miRNA expression file
```{r eval=TRUE, echo=TRUE}
miranda <- getInputSpecies("Epstein_Barr", threshold = 172)
uni_miRanda<-unique(miranda$V1)
uni_miRanda<-uni_miRanda[-28]#No expresion for this one miRNA in our dataset
```

## For interaction/synergistic miRNA-mRNA model

note all_coeff = TRUE only looks for interactions where both coefficients are negative for interaction mode or if you want to see positive interactions all_coeff = False might be useful.
```{r eval=TRUE, echo=TRUE}
miranda <- getInputSpecies("Epstein_Barr_Human", threshold = 140)#Use low thershod
testme <- runAllMirnaModels(uni_miRanda, DiffExpmRNA, DiffExpmiRNA, miranda, 0.1, nPar = 4, prob=0.95, method="fdr", all_coeff=TRUE, mode="inter")
```

## Output only interactions with a significant FDR

```{r eval=TRUE, echo=TRUE}
everything1 <- testme[sapply(testme, function(x) nrow(x$SigFDRGenes)) > 0]
```

## Extract correlation values/matrix from the model (R)

```{r eval=TRUE, echo=TRUE}
everything2<- rsquRes(everything1)
```

## Including correlation plots

```{r eval=TRUE, echo=TRUE, fig.width = 12, fig.height=5, out.width="90%", dpi=300, fig.align="center"}
DrawCorPlot(everything2, tl.cex = 0.3,cl.lim = c(-1, 0),tl.col = "black")
```

# Part2

Load files from test directory or you can laod individually and feed them in seperately in a list: list[(mRNA1,mRNA2,mRNA)]
```{r eval=TRUE, echo=TRUE}
files <- local({
  filenames <- list.files(path="test", pattern="^.*\\.txt$", full.names=T)
  files <- lapply(filenames, read.table, as.is=T, header=T, sep="\t")
  names(files) <- gsub("^.*/(.*)\\.txt$", "\\1", filenames)
  return(files)
})
```

## Get mRNAs

```{r eval=TRUE, echo=TRUE}
mrna_files <- files[grep("^mRNA", names(files))]
```

## Get mRNAs with particular fold change

```{r eval=TRUE, echo=TRUE}
mrna_files <- files[grep("^mRNA", names(files))]
mrna <- one2OneRnaMiRNA(mrna_files, pthreshold = 0.05)$foldchanges
```

## Get all miRNAs

```{r eval=TRUE, echo=TRUE}
mirna_files <- files[grep("^miRNA", names(files))]
mirna <- one2OneRnaMiRNA(mirna_files)$foldchanges
```

## Get mRNA miRNA correlation

```{r eval=TRUE, echo=TRUE}
corr_0 <- corMirnaRna(mrna, mirna,method="pearson")
```

## Make a background distribution correlation

```{r eval=TRUE, echo=TRUE, results=FALSE}
outs <- sampCorRnaMirna(mrna, mirna,method="pearson", Shrounds = 100, Srounds = 1000)
```

## Plot density plots 
Density plot for background and corrs in our data. Note grey is the background distribution and red is the actual data
```{r eval=TRUE, echo=TRUE, fig.width = 5, fig.height=5, out.width="80%", dpi=300, fig.align="center"}
mirRnaDensityCor(corr_0, outs)
```

## Get correlations below threshold

```{r eval=TRUE, echo=TRUE}
sig_corrs <- threshSig(corr_0, outs,pvalue = 0.05)
```

## Get mouse miranda data

```{r eval=TRUE, echo=TRUE}
miranda <- getInputSpecies("Mouse", threshold = 150)
```

## mRNA miRNA correlation heatmap 

Correlation heatmap for cor equal or less than -0.7. Note upperbound for heatmap should be always less than the correlation threshold
```{r eval=TRUE, echo=TRUE, fig.width = 7, fig.height=5, out.width="90%", dpi=300, fig.align="center"}
newcorr <- corMirnaRnaMiranda(mrna, mirna, -0.7, miranda)
mirRnaHeatmap(newcorr,upper_bound = -0.6)
```

## get intersection of miranda 

Get miranda intersection and significant miRNA and mRNA interactions and the plot it
```{r eval=TRUE, echo=TRUE, fig.width = 5, fig.height=5, out.width="80%", dpi=300, fig.align="center"}
results <- miRandaIntersect(sig_corrs, outs, mrna, mirna, miranda)
p<- mirRnaHeatmap(results$corr,upper_bound =-0.99)
p
```
# Part3

## Import data
```{r eval=TRUE, echo=TRUE}
files <- local({
  filenames <- list.files(path="test", pattern="^.*\\.txt$", full.names=T)
  files <- lapply(filenames, read.table, as.is=T, header=T, sep="\t")
  names(files) <- gsub("^.*/(.*)\\.txt$", "\\1", filenames)
  return(files)
})
```

## Only look for time point difference 0-2
```{r eval=TRUE, echo=TRUE}
mirna_files <- files[grep("^miRNA0_5", names(files))]
mrna_files <- files[grep("^mRNA0_5", names(files))]
```

## Get fold changes above thereshold
```{r eval=TRUE, echo=TRUE}
mrna <- one2OneRnaMiRNA(mrna_files, pthreshold = 0.05)$foldchanges
mirna <- one2OneRnaMiRNA(mirna_files)$foldchanges
```

## Get two time point interaction
```{r eval=TRUE, echo=TRUE}
inter0 <- twoTimePoint(mrna, mirna)
```

## Make correlation background distribution
```{r eval=TRUE, echo=TRUE, message=FALSE, results=FALSE}
outs <- twoTimePointSamp(mrna, mirna,Shrounds = 10 )
```

## Miranda data import
```{r eval=TRUE, echo=TRUE}
miranda <- getInputSpecies("Mouse", threshold = 140)
```

## get correlations below threshold
```{r eval=TRUE, echo=TRUE}
sig_InterR <- threshSigInter(inter0, outs)
```

## Miranda intersection with results
```{r eval=TRUE, echo=TRUE}
results <- miRandaIntersectInter(sig_InterR, outs, mrna, mirna, miranda)
```

## Make dataframe and plots
```{r eval=TRUE, echo=TRUE, fig.width = 7, fig.height=5, out.width="90%", dpi=300, fig.align="center"}
final_results <- finInterResult(results)
par(mar=c(4,4,2,1))
drawInterPlots(mrna,mirna,final_results)
```

## mRNA miRNA heatmap of miRNA mRNA FC differences 
Heatmap for p value significant miRNA mRNA fold change differences when compared to backgound
```{r eval=TRUE, echo=TRUE, fig.width = 5, fig.height=5, out.width="80%", dpi=300, fig.align="center"}
CorRes<-results$corrs
mirRnaHeatmapDiff(CorRes,upper_bound = 9.9)
```



